From 59f7f4a469a694c2da20e5fc8827c66c6645047a Mon Sep 17 00:00:00 2001
From: rusk <lordruskk@gmail.com>
Date: Fri, 25 Dec 2020 19:08:33 -0800
Subject: [PATCH] add more complex delimiter

---
 blocks.go | 10 +++++++---
 main.go   |  5 +++--
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/blocks.go b/blocks.go
index 24d422f..dd467ee 100644
--- a/blocks.go
+++ b/blocks.go
@@ -1,9 +1,13 @@
 package main
 
 var (
-	delim     = []byte(" ") // the delimiter that will be used
-	shell     = "sh"        // shell used
-	cmdstropt = "-c"        // command string opt for shell
+	delim = [3][]byte{ // the delimiter that will be used
+		{'['},         // bytes on the left side
+		[]byte("] ["), // bytes inbetween blocks
+		{']'},         // bytes on the right side
+	}
+	shell     = "sh" // shell used
+	cmdstropt = "-c" // command string opt for shell
 
 	blocks = []block{
 		{cmd: "recicon", upSig: 9},
diff --git a/main.go b/main.go
index 4f39d55..e758bee 100644
--- a/main.go
+++ b/main.go
@@ -100,12 +100,13 @@ func main() {
 	for range updateChan { // update bar on signal
 		for _, b := range blocks {
 			if barBytesArr[b.pos] != nil {
-				finalBytesBuffer.Write(delim)
+				finalBytesBuffer.Write(delim[1])
 				finalBytesBuffer.Write(barBytesArr[b.pos])
 			}
 		}
+		finalBytesBuffer.Write(delim[2])
 
-		finalBytes := bytes.TrimPrefix(finalBytesBuffer.Bytes(), delim)
+		finalBytes := append(delim[0], bytes.TrimPrefix(finalBytesBuffer.Bytes(), delim[1])[:]...)
 		xproto.ChangeProperty(x, xproto.PropModeReplace, root, xproto.AtomWmName, xproto.AtomString, 8, uint32(len(finalBytes)), finalBytes) // set the root window name
 		finalBytesBuffer.Reset()
 	}
-- 
2.29.2

